rswarp/diagnostics/ImpactDensity.py

```diff
@@ -81,52 +81,55 @@ def __init__(self, ax, ax_colorbar, scraper, top, w3d, interpolation='kde'):
             zoffset = 0.05 * np.amax([w3d.zmmin, w3d.zmmax])
             xoffset = 0.05 * np.amax([w3d.xmmin, w3d.xmmax])
-            ax.set_xlim((w3d.zmmin - zoffset) * self.scale[2], (w3d.zmmax + zoffset) * self.scale[2])
-            ax.set_ylim((w3d.xmmin - xoffset) * self.scale[0], (w3d.xmmax + xoffset) * self.scale[0])
+            if ax is not None:
+                ax.set_xlim((w3d.zmmin - zoffset) * self.scale[2], (w3d.zmmax + zoffset) * self.scale[2])
+                ax.set_ylim((w3d.xmmin - xoffset) * self.scale[0], (w3d.xmmax + xoffset) * self.scale[0])
 
             self.ax_colorbar = ax_colorbar
             self.cmap = cm.coolwarm
             self.normalization = Normalize
             self.cmap_normalization = None
         else:
-            self.ax = mlab.figure(1, bgcolor=(1, 1, 1), fgcolor=(0, 0, 0), size=(800, 600))
-            self.clf = mlab.clf()
+            self.normalization = Normalize
+            self.dy = w3d.dy
 
     def generate_plots_2d(self):
-        minS, maxS = maxint, 0
         scatter_plots = []
         for cond in self.conductors.itervalues():
-            print(cond)
-            for face in cond.generate_faces_2d():
-                x, z, s = face[0] * self.scale[0], \
-                             face[1] * self.scale[1], face[2] * e / self.time * 1e-4
-                print("m/m by face", np.min(s), np.max(s))
-                if 0 <= np.min(s) < minS:  # -1 value indicates no particle anywhere on face
-                    minS = np.min(s)
-                if np.max(s) > maxS:
-                    maxS = np.max(s)
-
+            face_data = self.generate_plot_data_for_faces_2d(cond)
+            for (x, z, s) in face_data:
                 if np.min(s) < 0.0:
                     scatter_plots.append(self.ax.scatter(z, x, c=s, s=1, linewidths=0, zorder=50))
                 else:
                     scatter_plots.append(self.ax.scatter(z, x, c=s, cmap=self.cmap, s=1, linewidths=0, zorder=50))
-        self.cmap_normalization = self.normalization(minS, maxS)
         ColorbarBase(self.ax_colorbar, cmap=self.cmap, norm=self.cmap_normalization)
 
+    def generate_plot_data_for_faces_2d(self, cond):
+        minS, maxS = maxint, 0
+        data = []
+        for face in cond.generate_faces_2d():
+            x, z, s = face[0] * self.scale[0], \
+                         face[1] * self.scale[1], face[2] * e / self.time * 1e-4
+            print("min/max by face", np.min(s), np.max(s))
+            if 0 <= np.min(s) < minS:  # -1 value indicates no particle anywhere on face
+                minS = np.min(s)
+            if np.max(s) > maxS:
+                maxS = np.max(s)
+
+            data.append((x, z, s))
+        self.cmap_normalization = self.normalization(minS, maxS)
+        return data
+
     def generate_plots_3d(self):
+        self.ax = mlab.figure(1, bgcolor=(1, 1, 1), fgcolor=(0, 0, 0), size=(800, 600))
+        self.clf = mlab.clf()
+
         minS, maxS = maxint, 0
```

Comment:
These variables are still being initialized and used outside of `generate_plot_data_for_faces_3d` but the correct values aren't being passed out of `generate_plot_data_for_faces_3d` which causes VTK to throw an error when making the color table.

Either setting their values in  `generate_plot_data_for_faces_3d` and passing it out or setting them during the loop over `face_data` in `generate_plots_3d` is fine with me. Whichever works better for your purposes.